### ISSUE TITLE
(Intermediate) If running the docker image in foreground, sending a kill signal doesn't stop the container

### ISSUE THREAD

**Issue by rishabhpoddar** [supertokens/supertokens-core#28]
To recreate the problem that exists today:
- run `docker pull supertokens/supertokens-mysql:2.4`
- Start the docker image in foreground: `docker run -p 3567:3567 supertokens/supertokens-mysql:2.4`
- Wait for it to start.. It should display `Started SuperTokens on 0.0.0.0:3567 with PID: <some PID>`
- Now send the kill signal (Ctrl + C / Z on a Mac)

At this point, the Java process in the docker process should stop and the docker process itself should quit. However, that does not happen.

The docker file for the above image can be found in [this repo](https://github.com/supertokens/supertokens-docker-mysql). Please feel free to fork that as well.

As some extra info:
- if not using docker (manually installing it on your system), sending the kill signal does stop the running process.
- In the Java code (this repo), in [Main.java](https://github.com/supertokens/supertokens-core/blob/master/src/main/java/io/supertokens/Main.java), there is [a function](https://github.com/supertokens/supertokens-core/blob/master/src/main/java/io/supertokens/Main.java#L354) that registers a shutdown hook.

**Comment by jz543fm** [comment]
Hi, can you assign this to me and I will check this out, and could you send me the proper information?

**Comment by rishabhpoddar** [comment]
@jz543fm assigned! Thanks for that! I have added details about this issue on top. Please feel free to ask more questions.

**Comment by jz543fm** [comment]
try` $ docker run -d -it -p 3567:3567 supertokens/supertokens-mysql:2.4`
and then

$ docker logs <Container ID or NAME>

**Comment by rishabhpoddar** [comment]
I tried that. Though you are running it in the background with the `-d` option. If that's not given, and the kill signal is sent, then the problem mentioned in this issue arises. Please feel free to clarify your last comment if I misunderstood :)

**Comment by jz543fm** [comment]
Yes, but you are able to kill process via docker kill, or docker stop
container, docker inspect container to check config or docker stop and then
docker rm, to stop container and remove container, you Can access the CLI
and run commands because when I tried It on Linux and Windows I could not
use CTRL c or CTRL z, Also you Can check logs Via docker logs command

Dňa ut 29. 9. 2020, 22:42 Rishabh Poddar <notifications@github.com>
napísal(a):

**Comment by rishabhpoddar** [comment]
Yea you can kill the process via `docker stop`. But this issue is about a different problem, which still exists. Have you not been able to replicate it by following the instructions in the first comment?

**Comment by Fabricio20** [comment]
Hey there, looking at this issue, seems like the cause may lie on the bash wrapper/entrypoint. Sending a SIGTERM to  PID 1 won't cause anything, but sending SIGTERM to the JVM below it will, the signal on the docker CLI is sent to PID 1 so that explains why nothing happens.

**Comment by rishabhpoddar** [comment]
@Fabricio20 , thanks! Would you be able to contribute the change or show me what to change with an example please?

**Comment by Fabricio20** [comment]
@rishabhpoddar Currently looking into the issue, seems like a normal bash behaviour. Any special parameters needed to build the image over on the other repository?

**Comment by radiantly** [comment]
Usually this can be fixed by adding `--init` flag to the docker run command since bash is replaced when using exec.

**Comment by rishabhpoddar** [comment]
@Fabricio20 
> Any special parameters needed to build the image over on the other repository?

Assuming you are using the [supertokens-docker-mysql](https://github.com/supertokens/supertokens-docker-mysql) repo, there is nothing special needed to build the docker image (if you are not modifying `supertokens-core`). You can simply use something like `docker build -t supertokens-mysql .`

**Comment by rishabhpoddar** [comment]
@radiantly , we would ideally not want to enforce adding an extra flag for the user, unless this is common practice. However, I have never used the `--init` flag before in a `docker run` command, so not sure. Please correct me if I am wrong :)

**Comment by rishabhpoddar** [comment]
@Fabricio20 , I tried to test out the change you made, however, when I install SuperTokens, the resulting CLI script has `null` instead of what's in [this file](https://github.com/Fabricio20/supertokens-core/blob/2eb333c9870756a7efcf2ac59e79f6f20fd5201c/cli/src/main/resources/install-linux.sh)

**Comment by rishabhpoddar** [comment]
There is a problem with the linux script in which if one does not use docker, it runs all commands (like `supertokens list`) in the background. So I propose that the script be modified to check if it's running in a docker container and only then running it in the background like so:
```
#!/bin/bash
ST_INSTALL_LOC=$ST_INSTALL_LOC
if [ -f /.dockerenv ]; then
    trap 'kill -TERM $PID' TERM INT
    "${ST_INSTALL_LOC}"jre/bin/java -classpath "${ST_INSTALL_LOC}cli/*" io.supertokens.cli.Main false "${ST_INSTALL_LOC}" $@ &
    PID=$!
    wait $PID
else
    "${ST_INSTALL_LOC}"jre/bin/java -classpath "${ST_INSTALL_LOC}cli/*" io.supertokens.cli.Main false "${ST_INSTALL_LOC}" $@
fi
```

A few notes:
- I have removed the `if ... && [ $1 == update ]` clause since that is not needed anymore.
- The `if [ -f /.dockerenv ]; then` checks if the script is running in the docker container, however, I'm not sure if this is the recommended way of doing so.
- If one runs the docker container in interactive mode, and then starts supertokens, it causes all sorts of problems.
- The Windows version of the script can remain exactly the same since the docker image is never going to be on Windows OS.

**Comment by Fabricio20** [comment]
Hey, answering your questions:

> resulting CLI script has null instead of what's in this file

I'll look into it later today, it should automatically be shaded by gradle however.

> The if `[ -f /.dockerenv ];` then checks if the script is running in the docker container, however, I'm not sure if this is the recommended way of doing so.

According to [this](https://stackoverflow.com/a/41559867/4592841) SO answer it seems like the preferred way would be to check cgroup since `.dockerenv`  could be removed at any moment (read about [.dockerinit](https://github.com/docker/docker/issues/18355)).

> If one runs the docker container in interactive mode, and then starts supertokens, it causes all sorts of problems.

What kind of issues are you experiencing? This should trap the sigs sent to bash and recreate them as a kill command for the child's pid. (So even if starting manually this should have the same behaviour).

> The Windows version of the script can remain exactly the same since the docker image is never going to be on Windows OS.

Windows script shouldn't have been changed, just repackaged it on resources for the sake of consistency.

**Comment by rishabhpoddar** [comment]
> I'll look into it later today, it should automatically be shaded by gradle however.

Thanks! When building, there is a separate CLI jar gets generated based on [this build.gradle file](https://github.com/supertokens/supertokens-core/blob/master/cli/build.gradle).

> What kind of issues are you experiencing?

The issue is that after running a command like `supertokens --help` (inside the docker container), it gives me control of the shell and then prints out the output, which means that I can type (and execute) other commands while the output of `supertokens --help` is being written to the console at the same time.

**Comment by Fabricio20** [comment]
I've pushed the changes to my branch in an attempt to fix the resource loading issue. I managed to build the project and confirmed the `cli.jar` has the resources inside it.

I haven't been able to build the docker image however, seems to be stuck at `Preparing test environment` or something very similar on `./startTestingEnv --wait`. I pushed the changes you suggested to the linux image which should prevent issues on the default/non-docker installation.

As for the shell issue.. that shouldn't happen as it triggers a `wait` on the child PID (unless your sh is ignoring it?).

**Comment by rishabhpoddar** [comment]
Thank you @Fabricio20 . I'll test with the docker image once again and let you know what happens :)

**Comment by igoose1** [comment]
I've used 4.6 and 6.0 from docker and I see the same behavior: CTRL + C doesn't kill the process, `docker stop` does though. @rishabhpoddar, did it become normal after a few years? I didn't find other issues related to the problem.

Unstoppable command: `docker run -p "3567:3567" -e API_KEYS=... registry.supertokens.io/supertokens/supertokens-postgresql:6.0`

OS: Ubuntu 22.10.

**Comment by igoose1** [comment]
Related: #49.

**Comment by rishabhpoddar** [comment]
@igoose1 can you give all the options given to the docker run command? (other than the env vars)

**Comment by igoose1** [comment]
@rishabhpoddar, actually the entire command I provided was used. I didn't specify arguments. If I understand it correctly, it means docker sets `--foreground` for me.

I debugged [docker-entrypoint.sh](https://github.com/supertokens/supertokens-docker-postgresql/blob/b7cf84480eeac5e8fd38ece562c9f8bcb92fa744/docker-entrypoint.sh#L335C1-L346C3) on my own. Look at my comments:

```sh
# [...]
# 
# check if no options has been passed to docker run
if [[ "$@" == "supertokens start" ]]
then
    # ISSUE RELATED: this was executed!
    set -- "$@" --foreground
fi


# If container is started as root user, restart as dedicated supertokens user
if [ "$(id -u)" = "0" ] && [ "$1" = 'supertokens' ]; then
    # ISSUE RELATED: this was executed!
    exec gosu supertokens "$@"
else
    exec "$@"
fi
```

**Comment by rishabhpoddar** [comment]
Thanks @igoose1. Will try and solve this is and when we have time.
